### PR TITLE
Fix pyspnego

### PR DESCRIPTION
**Pull request by csernazs**
###### Description of changes

pyspnego is broken as its `krb5` dependency in the nix expression is specified incorrectly. For example, when it is used by `requests-kerberos` python library, it raises `ImportError` exception.

Example test which shows the issue:
```python
import requests
from pytest_httpserver import HTTPServer
from requests_kerberos import HTTPKerberosAuth


def test_spnego(httpserver: HTTPServer):
    httpserver.expect_request("/auth").respond_with_data(status=401, headers={"WWW-Authenticate": "Negotiate"})
    resp = requests.get(httpserver.url_for("/auth"), auth=HTTPKerberosAuth())

    assert resp.status_code == 401
```

With current nixpkgs master it produces:
```
[...]
        if not HAS_GSSAPI:
>           raise ImportError("GSSAPIProxy requires the Python gssapi library: %s" % GSSAPI_IMP_ERR)
E           ImportError: GSSAPIProxy requires the Python gssapi library: No module named 'krb5'
```

The problem is that `pyspnego` depends on the `pykrb5` python library and not on the `krb5` derivation as specified (which contains the C library, and the related toolchain).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC: @fabaff (pyspnego) @catern (requests-kerberos)

### CODE REVIEW

File: pkgs/development/python-modules/pykrb5/default.nix [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+buildPythonPackage rec {
+  pname = "krb5";
+  version = "0.4.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-brDpHgRNN8s0yEuDExD0pIlR8pAtTrtluXy6xf8bImQ=";
+  };
+
+  nativeBuildInputs = [ krb5Full ]; # for krb5-config
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pykrb5/default.nix:12]:
```suggestion
  nativeBuildInputs = [
    krb5.dev # for krb5-config
  ];
```
would that work?

File: pkgs/development/python-modules/pykrb5/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  buildInputs = [ cython ];
+  checkInputs = [ pytestCheckHook k5test which ];
+  doCheck = true;
+
+  meta = with lib; {
+    description = "Kerberos API bindings for Python";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pykrb5/default.nix:19]: homepage should be set

File: pkgs/development/python-modules/pykrb5/default.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-brDpHgRNN8s0yEuDExD0pIlR8pAtTrtluXy6xf8bImQ=";
+  };
+
+  nativeBuildInputs = [ krb5Full ]; # for krb5-config
+
+  buildInputs = [ cython ];
+  checkInputs = [ pytestCheckHook k5test which ];
+  doCheck = true;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pykrb5/default.nix:16]:
```suggestion
```
that is the default

File: pkgs/development/python-modules/k5test/fix-paths.patch [hunk 4]
```diff
@@ -26,97 +26,100 @@ index e289fac..e2f486a 100644
...
+         _KRB_VERSION = _KRB_VERSION.split(" ")[-1].split(".")
  
      def make_ext_test(func):
+diff --git a/setup.py b/setup.py
+index e5909f6..0132c9a 100755
+--- a/setup.py
++++ b/setup.py
+@@ -1,4 +1,4 @@
+-#!/usr/bin/env python
++#!@krb5FullDev@/bin/env python
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/k5test/fix-paths.patch:122]: That should be patched with patchShebangs

File: pkgs/development/python-modules/k5test/fix-paths.patch [hunk 5]
```diff
@@ -26,97 +26,100 @@ index e289fac..e2f486a 100644
...
-index 3c64b9e..59da1ab 100644
+index caa8315..5bcf0bf 100644
 --- a/k5test/unit.py
 +++ b/k5test/unit.py
-@@ -38,7 +38,7 @@ _KRB_VERSION = None
+@@ -39,7 +39,7 @@ _KRB_VERSION = None
  def krb_minversion_test(target_version, problem, provider=None):
      global _KRB_VERSION
      if _KRB_VERSION is None:
 -        _KRB_VERSION = _utils.get_output("krb5-config --version")
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/k5test/fix-paths.patch:111]: Wouldn't it work when taking it from PATH?

File: pkgs/development/python-modules/k5test/fix-paths.patch [hunk 6]
```diff
@@ -26,97 +26,100 @@ index e289fac..e2f486a 100644
...
--            ('_kinit', 'kinit', '/usr/bin/kinit'),
--            ('_klist', 'klist', '/usr/bin/klist'),
-+            ('kdb5_util', 'kdb5_util', '@krb5Full@/bin/kdb5_util'),
-+            ('krb5kdc', 'krb5kdc', '@krb5Full@/bin/kdb5kdc'),
-+            ('kadmin', 'kadmin', '@krb5Full@/bin/admin'),
-+            ('kadmin_local', 'kadmin.local', '@krb5Full@/bin/kadmin.local'),
-+            ('kadmind', 'kadmind', '@krb5Full@/bin/kadmind'),
-+            ('kprop', 'kprop', '@krb5Full@/bin/kprop'),
-+            ('_kinit', 'kinit', '@krb5Full@/bin/kinit'),
-+            ('_klist', 'klist', '@krb5Full@/bin/klist'),
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/k5test/fix-paths.patch:88]: I would prefer taking those from PATH, too when possible.

File: pkgs/development/python-modules/pykrb5/default.nix (hunk 1 above)
Comment (csernazs) [reply pkgs/development/python-modules/pykrb5/default.nix:12]: yes, fixed

File: pkgs/development/python-modules/k5test/fix-paths.patch (hunk 4 above)
Comment (csernazs) [reply pkgs/development/python-modules/k5test/fix-paths.patch:122]: I think this was done by mistake, I removed this patching. I think `patchShebangs` is not needed for setup.py as it is run by python explicitly.

Do you agree? If not, I can patch it, sure.

File: pkgs/development/python-modules/k5test/fix-paths.patch (hunk 5 above)
Comment (csernazs) [reply pkgs/development/python-modules/k5test/fix-paths.patch:111]: This is not the test code for k5test, for that, we could set `PATH` during the build I think. It is the python module which is used by other programs for testing.
If we require that this should be used from `PATH`, we will require the other programs to put the krb5 tools (such as `krb5-config`) to their `PATH` envvar during the check phase.

Is this what you wanted?

I think k5test is opaque from the users point of view.

File: pkgs/development/python-modules/k5test/fix-paths.patch (hunk 6 above)
Comment (csernazs) [reply pkgs/development/python-modules/k5test/fix-paths.patch:88]: See above.

File: pkgs/development/python-modules/k5test/fix-paths.patch (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/k5test/fix-paths.patch]: if it works, it works

**Comment by lukebfox** [comment]
+1 for this fix thanks

**Comment by csernazs** [comment]
@SuperSandro2000 @mweinelt could you please take a look at this PR again? Thanks!

**Comment by jljusten** [comment]
I think I'm also seeing this with the `requests-kerberos` python module. (Or maybe `gssapi`?)

**Comment by de11n** [comment]
I believe this was fixed by #333058.

**Comment by SuperSandro2000** [comment]
Great, closing this as stale and replaced

**Pull request closed without merging**